### PR TITLE
Generic model level

### PR DESCRIPTION
**Pull request by DashDeipayan**
created apis for creating, removing generic levels for items

### CODE REVIEW

File: routes/levels.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+const express = require("express");
+const { addLevel, deleteLevel, getAllLevels } = require("../controllers/levels");
+const { validateLevelBody } = require("../middlewares/validators/levels");
+
+const router = express.Router();
+
+router.post("/", validateLevelBody, addLevel);
+router.delete("/:levelid", deleteLevel);
```
Comment (vinayak-trivedi) [inline routes/levels.js:8]: we do not want everyone to create levels, authorize only superuser to do that

File: controllers/levels.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+/**
+ * Creates a level
+ *
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+
+const addLevel = async (req, res) => {
+  try {
+    const { id, levelData } = await LevelModel.addLevel(req.body);
```
Comment (vinayak-trivedi) [inline controllers/levels.js:12]: can we get the data of the person who created the tag using authenticate middleware?

File: models/levels.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ *
+ * @param id { String }: id of the level to be deleted
+ * @returns {Promise<tagModel|String>}
+ */
+
+const deleteLevel = async (id) => {
+  try {
+    await levelModel.doc(id).delete();
```
Comment (vinayak-trivedi) [inline models/levels.js:28]: can we destructure `id` here because I saw that you are sending req.body in the argument there

**Review by vinayak-trivedi — CHANGES_REQUESTED** [review]
Great work @DashDeipayan 
suggesting some changes please check

File: routes/levels.js (hunk 1 above)
Comment (DashDeipayan) [reply routes/levels.js]: done

File: controllers/levels.js (hunk 2 above)
Comment (DashDeipayan) [reply controllers/levels.js:13]: done

File: models/levels.js (hunk 3 above)
Comment (DashDeipayan) [reply models/levels.js:28]: fixed it.

**Review by vinayak-trivedi — DISMISSED** [review]
LGTM 👍

**Review by alishvagadiya — APPROVED** [review]

File: controllers/levels.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+/**
+ * Creates a level
+ *
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+
+const addLevel = async (req, res) => {
+  try {
+    const { id, levelData } = await LevelModel.addLevel({ ...req.body, createdby: req.userData.id, date: new Date() });
```
Comment (ivinayakg) [inline controllers/levels.js:12]: epoch not in the date

**Review by ivinayakg — CHANGES_REQUESTED** [review]

File: models/levels.js (hunk 3 above)
Comment (alishvagadiya) [reply models/levels.js:29]: Why do we need an id in return response?

We are deleting data, which means we don't require any more.
it should only success message ``` Data deleted successfully```

File: models/levels.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  }
+};
+
+/**
+ *
+ * @returns {Promise<tagModel|Array>}
+ */
+
+const getAllLevels = async () => {
+  const data = await levelModel.get();
```
Comment (alishvagadiya) [inline models/levels.js:43]: same comment as above for selecting data.

File: middlewares/validators/levels.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+const Joi = require("joi");
+
+const validateLevelBody = async (req, res, next) => {
+  const schema = Joi.object({
+    name: Joi.string().trim().required(),
+    levelnumber: Joi.number().integer().min(0).required(),
+  });
+
+  try {
```
Comment (alishvagadiya) [inline middlewares/validators/levels.js:9]: Add Joi part in a try block.

File: controllers/levels.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Collects and returns all levels
+ *
+ * @param res {Object} - Express response object
+ */
+
+const getAllLevels = async (req, res) => {
+  try {
+    const { allLevels } = await LevelModel.getAllLevels();
+    return res.json({
```
Comment (alishvagadiya) [inline controllers/levels.js:58]: Return response formate is changing at all control functions. Make it consistent.

File: models/levels.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    throw err;
+  }
+};
+
+/**
+ *
+ * @returns {Promise<tagModel|Array>}
+ */
+
+const getAllLevels = async () => {
```
Comment (alishvagadiya) [inline models/levels.js:42]: Add try-catch block

**Review by alishvagadiya — CHANGES_REQUESTED** [review]

File: controllers/levels.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+
+const deleteLevel = async (req, res) => {
+  try {
+    const { id } = await LevelModel.deleteLevel(req.params.levelid);
+    return res.json({
+      message: "Level Deleted successfully!",
+      id,
```
[the remark this answers is not in the document]
Comment (DashDeipayan) [reply controllers/levels.js:41]: Just as a confirmation, no tangible purpose

File: middlewares/validators/levels.js (hunk 6 above)
Comment (DashDeipayan) [reply middlewares/validators/levels.js:9]: all other files have validation in this format.

File: models/levels.js (hunk 3 above)
Comment (vinayak-trivedi) [reply models/levels.js:29]: makes sense let me do it

File: models/levels.js (hunk 5 above)
Comment (vinayak-trivedi) [reply models/levels.js]: will be done in the latest commit

**Review by alishvagadiya — APPROVED** [review]

**Pull request merged**